lib/api/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+    pub use wasmer_runtime_core::table::Table;
+    pub use wasmer_runtime_core::types::{
+        FuncSig, GlobalDescriptor, MemoryDescriptor, TableDescriptor, Type, Value,
+    };
+}
+
+pub mod import {
+    //! Types and functions for Wasm imports.
+    pub use wasmer_runtime_core::module::{Import, ImportType};
+    pub use wasmer_runtime_core::{func, imports};
+}
+
+pub mod export {
+    //! Types and functions for Wasm exports.
+    pub use wasmer_runtime_core::module::{ExportDescriptor, ExportKind};
+}
+
+pub mod units {
+    //! Various unit types.
+    pub use wasmer_runtime_core::units::{Bytes, Pages};
+}
+
+pub mod types {
+    //! Types used in the Wasm runtime and conversion functions.
+    pub use wasmer_runtime_core::types::*;
+}
+
+pub mod error {
+    //! Various error types returned by Wasmer APIs.
+    pub use wasmer_runtime_core::error::*;
+}
+
+pub use prelude::*;
+
+/// Idea for generic trait; consider rename; it will need to be moved somewhere else
+pub trait CompiledModule {
+    fn new(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_binary(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_binary_unchecked(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_file(file: impl AsRef<std::path::Path>) -> error::CompileResult<Module>;
+
+    fn validate(bytes: impl AsRef<[u8]>) -> error::CompileResult<()>;
+}
+
+use wasmer_runtime_core::backend::Compiler;
+
+/// Copied from runtime core; TODO: figure out what we want to do here
+pub fn default_compiler() -> impl Compiler {
+    #[cfg(any(
+        all(
+            feature = "default-backend-llvm",
+            not(feature = "docs"),
+            any(
+                feature = "default-backend-cranelift",
+                feature = "default-backend-singlepass"
+            )
+        ),
+        all(
+            not(feature = "docs"),
+            feature = "default-backend-cranelift",
+            feature = "default-backend-singlepass"
+        )
+    ))]
+    compile_error!(
+        "The `default-backend-X` features are mutually exclusive.  Please choose just one"
+    );
+
+    #[cfg(all(feature = "default-backend-llvm", not(feature = "docs")))]
+    use wasmer_llvm_backend::LLVMCompiler as DefaultCompiler;
+
+    #[cfg(all(feature = "default-backend-singlepass", not(feature = "docs")))]
+    use wasmer_singlepass_backend::SinglePassCompiler as DefaultCompiler;
+
+    #[cfg(any(feature = "default-backend-cranelift", feature = "docs"))]
+    use wasmer_clif_backend::CraneliftCompiler as DefaultCompiler;
+
+    DefaultCompiler::new()
+}
+
+// this implementation should be moved
+impl CompiledModule for Module {
+    fn new(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
+        let bytes = bytes.as_ref();
+        wasmer_runtime_core::compile_with(bytes, &default_compiler())
+    }
+
+    fn from_binary(_bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
```

Comment:
I think that should be just a `Result` ?

Reply:
What did you have in mind for the error type?

Reply:
We can use `anyhow::Error` :)

Reply:
Hmm, I'll have to research it more but my experience with third-party error crates is that they provide a nice way to bundle all your errors together but it comes at the cost of making the errors essentially impossible to handle in code.  Perhaps `anyhow` has a solution for that, but I'd be hesitant to go behind basic `enums` and maybe some `structs` if we expect users to able to handle the errors.

Reply:
From https://github.com/dtolnay/anyhow#comparison-to-thiserror :

> Use Anyhow if you don't care what error type your functions return, you just want it to be easy. This is common in application code. Use thiserror if you are a library that wants to design your own dedicated error type(s) so that on failures the caller gets exactly the information that you choose.

This confirms my assumption that the library is for when you don't want to handle the errors and shouldn't be used in our public API.  We can use `thiserror` (by the same author) if we want a library to help us with error types, but I think it might be simpler to just be explicit with an enum.